orttraining/orttraining/training_api/module.cc

```diff
@@ -103,39 +105,45 @@ Module::Module(const std::string& train_model_path_or_bytes,
       ORT_ENFORCE(target_device == *(it->device), "Inconsistent device requirements found for input: ", param_name);
     }
 
-    // Create parameter value copy with corresponding device user sets the session on.
-    // We did not re-use the data even CPU tensor is needed.
     // TODO(pengwa): consider whether we should alloc contiguous buffer for parameters or gradients.
-    OrtValue target_ortvalue;
-    auto allocator = train_sess_state.GetAllocator(target_device);
-    ORT_ENFORCE(allocator != nullptr);
-
-    Tensor::InitOrtValue(source_tensor.DataType(),
-                         source_tensor.Shape(),
-                         allocator, target_ortvalue);
-    Tensor* target_tensor_ptr = target_ortvalue.GetMutable<Tensor>();
-    ORT_THROW_IF_ERROR(train_sess_state.GetDataTransferMgr().CopyTensor(source_tensor, *target_tensor_ptr));
+    // Copy ortvalue buffer from CPU to target_device for this "param_name" (based on graph partitioning)
+    // Only copies data if target device is not the same as the current device the buffer is placed on
+    auto params_iter = named_parameters_.find(param_name);
```

Comment:
we are finding the param_name twice( in the enforce above line 97), can be merged to 1.